### PR TITLE
Test run on more program ids with new hst pipeline under hst dir

### PR DESCRIPTION
**Pull request by juzen2003**
- Replace https://github.com/SETI/pds-hst-pipeline/pull/71

### CODE REVIEW

**Comment by juzen2003** [comment]
Update based on Rob's feedback on the task queue manager:

        - Modify the architecture to spawn all the subprocesses at a center location (inside run_pipeline function). Each subtask would queue its successor, but not run it. It would then just exit. The top-level pipeline_run would be in a loop waiting for a subprocess slot to become open, and then read the next task from the table (for any program id) and run it.  
        - Update setup document with commands to run the program.
        - Update & clean up the code based on Rob's feedback in queue_manager/__init__.py, queue_manager/config.py and queue_manager/task_queue_db.py
        - Change all 'mast' and 'Mast' to 'MAST' for files under HST/
        - Re-org files, move the old files under archive/
        - Update requirements.txt to work with python 3.11
        - Remove ending periods in logger messages.
        - Turn off sqlalchemy terminal console messages.
        - Fix pipeline programs file names in the ArgumentParser's descriptions.
        - Add max subprocesses and max allowed time to command line options.
        - Remove pds-tools from PYTHONPATH, use pip-installable versions of julian, textkernel, pdslogger and pdstemplate.
        - Setup requirements.in using pip-compile from pip-tools to auto generate requirements.txt 
        - Add a step in `retrieve_hst_visit.py` to remove the corresponding trl file when downloading MAST files has failed. That way in query_hst_products task, we know that the MAST downloading for a specific propsal id & visit is not done, and the files of that visit will be treated as new files waiting to be downloaded in the next task.
        - Replace task number with task name string, and create TASK_INFO dictionary to replace the TASK_*_MAPPING constants
        
Example commands:
`python pipeline/pipeline_run.py --prog-id 07885 7315 08659 06328 8218 13736 05167 10341 14930 16310 9256 7309 15406 10095 6621 5392 6736 9985 7239 10156 --max_subproc 30 --max_time 1860`

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
This is just an initial set of comments before I run the pipeline.

- Generally speaking, command line arguments shouldn't have underscores in them. Just `--proposal-ids` instead of `--proposal_ids`.

**Comment by juzen2003** [comment]
- Update based on feedback, here are example commands:
    - Run with the full ids (pre-fetch from MAST with True moving target flag):
        - `python pipeline/pipeline_run.py`
    - Query MAST with True moving target flag to get the latest ids, and then run pipeline with them:
        - `python pipeline/pipeline_run.py --get-ids`
    - Run with one proposal id:
        - `python pipeline/pipeline_run.py --prog-id 07885`
    - Run with multiple proposal ids:
        - `python pipeline/pipeline_run.py --prog-id 13736 05167 10341 14930 06679`
    - Run with specific number of subprocesses and max allowed running time for each task (in sec):
        - `python pipeline/pipeline_run.py --prog-id 07885 13736 --max-subproc 30 --max-time 1860`

File: HST/queue_manager/config.py [hunk 1]
```diff
@@ -18,36 +18,101 @@
 DB_PATH = f'{HST_DIR["pipeline"]}/task_queue.db'
 DB_URI = f'sqlite:///{DB_PATH}'
 
-# max allowed suborprocess time in seconds
-MAX_ALLOWED_TIME = 60 * 60
+# max allowed subprocess time in seconds
+MAX_ALLOWED_TIME = 60 * 30
 # max number of subprocesses to run at a time for one hst pipeline on a proposal id
-MAX_SUBPROCESS_CNT = 10
+MAX_SUBPROCESS_CNT = 20
```
[the remark this answers is not in the document]
Comment (juzen2003) [reply HST/queue_manager/config.py:25]: Yes, this is the max number of the total subprocesses for the pipeline run with all ids combined. The comment was wrong, and it's updated now.

File: HST/organize_files.py [hunk 2]
```diff
@@ -36,3 +36,4 @@ def organize_files_from_staging_to_bundles(proposal_id, logger):
                 os.makedirs(bundles_prod_dir, exist_ok=True)
                 logger.info(f'Move {dir} from staging to bundles directory')
                 shutil.copytree(staging_prod_dir, bundles_prod_dir, dirs_exist_ok=True)
+                shutil.rmtree(staging_prod_dir)
```
[the remark this answers is not in the document]
Comment (juzen2003) [reply HST/organize_files.py:56]: The code is updated, and now we will clean up the staging folder after the whole pipeline is done.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
I also notice these uses of `rmtree` throughout the code. All should be looked at carefully:
```
HST/retrieve_hst_visit.py:        shutil.rmtree(files_dir)
HST/query_hst_products.py:                shutil.rmtree(dir_path)
HST/query_hst_products.py:    shutil.rmtree(staging_dir + '/mastDownload')
HST/organize_files.py:                shutil.rmtree(staging_prod_dir)
HST/tests/test_get_program_info.py:        shutil.rmtree(self.download_dir)
HST/tests/test_retrieve_hst_visit.py:            shutil.rmtree(testing_dir)
HST/tests/test_label_creations.py:            shutil.rmtree(testing_dir)
```

**Comment by juzen2003** [comment]
- Updates based on comments:
    - Remove `*pdscache.py` and `*translator.py` in `.coveragerc`
    - Remove `--prog-id` in the command line arguments, use `--proposal-id` and
    `--proposal-ids` correspondingly
    - Update command line argument `--visit-li` to `--visit-list` to be more clear and remove its shorter version `--vi`
    - Update `setup.md` due to command line arguments changed
    - For `shutil.rmtree`:
```
HST/retrieve_hst_visit.py:              shutil.rmtree(files_dir) --> this is to removed the mastDownload dir when there is a failure during MAST downloading
HST/query_hst_products.py:              shutil.rmtree(dir_path) --> this is to removed the mastDownload dir when there is a failure during MAST downloading
HST/query_hst_products.py:              shutil.rmtree(staging_dir + '/mastDownload') --> Move to final clean up function clean_up_staging_dir 
HST/organize_files.py:                  shutil.rmtree(staging_prod_dir) --> this is in the final clean up function clean_up_staging_dir
HST/tests/test_get_program_info.py:     shutil.rmtree(self.download_dir) --> use in teardown function for testing, remove the download dir used in the test
HST/tests/test_retrieve_hst_visit.py:   shutil.rmtree(testing_dir) --> use in teardown function for testing, remove the download dir used in the test
HST/tests/test_label_creations.py:      shutil.rmtree(testing_dir) --> use in teardown function for testing, remove the download dir used in the test
```


Remaining issues:
- Label errors from the validation tool `https://github.com/SETI/rms-hst-pipeline/issues/56` (Looking at it now)
- Issues related to label versioning, details in `https://github.com/SETI/rms-hst-pipeline/issues/53`
- Issues happen during label creation caused by missing files from MAST (currently we log the error and continue the pipeline), details in: 
```
https://github.com/SETI/rms-hst-pipeline/issues/65
https://github.com/SETI/rms-hst-pipeline/issues/66
```
- Issue related to undetermined wavelength, expand FILTER_EXCEPTIONS or FILTER_REGEX? Details in `https://github.com/SETI/rms-hst-pipeline/issues/70`
- Issue related to unrecognized EXTNAME, expand dictionaries of DESCRIPTIONS or EXTRA_DESCRIPTIONS? Details in `https://github.com/SETI/rms-hst-pipeline/issues/67`
- Put SPICE kernel to the final bundle, details in:
```
https://github.com/SETI/rms-hst-pipeline/issues/78
https://github.com/SETI/rms-hst-pipeline/issues/72
```
- Add labels for browse products, details in `https://github.com/SETI/rms-hst-pipeline/issues/73`
- Corner case: handle the case for a target body with multiple names & types, details in `https://github.com/SETI/rms-hst-pipeline/issues/60`

**Review by rfrenchseti — COMMENTED** [review]
One small comment: Since we have `--proposal-ids` shouldn't it be `--visits` instead of `--visit-list`?

**Comment by juzen2003** [comment]
Change command line argument "--visit-list" to "--visits"

**Comment by juzen2003** [comment]
- Fixed some errors caused by validator.

Modifications:
- In PRODUCT_LABEL.xml
    - Change `hst:moving_target_keyword` to `hst:moving_target_keywords`
    - Remove attribute `nilReason` in `hst:gain_setting` tag (attribute is not allowed)
    - Remove attribute `unit` in `hst:plate_scale` tag (attribute is not allowed)
    - Remove attribute `nilReason` in `hst:spectral_resolution` tag (attribute is not allowed)
    - Remove attribute `nilReason` in `hst:center_filter_wavelength` tag (attribute is not allowed)
    - Remove attribute `nilReason` in `hst:bandwidth` tag (attribute is not allowed)
    - Add `unit="byte"` to `record_length`, `field_location`, and `field_location` tags
- In hst_dictionary_support.py
    - Change `hst:observation_type` from `SPECTROSCOPIC` back to `SPECTROGRAPHIC`
    - Change `Array_1D` to `Array`
- In PRODUCT_LABEL.xml
    - Change `File_Area` to `File_Area_Ancillary` tag if processing level is "Ancillary"
- In all labels templates, keep `logical_identifier` tag and its content in one line to avoid validator complaining about the unexpected carriage returns



Issues that needs to be reviewed/discussed:

- `hst:gain_setting`, `hst:spectral_resolution`, `hst:center_filter_wavelength`, and `hst:bandwidth` values need to match this pattern `r'(\+|-)?([0-9]+(\.[0-9]*)?|\.[0-9]+)([Ee](\+|-)?[0-9]+)?|[^aFIN,]* '`, what should we put if it's not applicable? Remove the tag? (empty string is not a valid value)
- No child element allowed for these tags" `hst:visit_id` & `hst:Processing_Parameters`
- For `hst:observation_type`, it must exists, and the validator requires the value to be one of these: 'IMAGING', 'SPECTROGRAPHIC', 'TIME-SERIES', so 'UNK' will fail the validator.
- For data class, these are the available values: `{Array, Array_2D, Array_2D_Image, Array_2D_Map, Array_2D_Spectrum, Array_3D, Array_3D_Image, Array_3D_Movie, Array_3D_Spectrum, Checksum_Manifest, Encoded_Header, Encoded_Image, Header, Stream_Text, Table_Binary, Table_Character, Table_Delimited}`. Should we modify `DATA_CLASS_TO_NOUN` in `hdu_data_descriptions.py`? What value should we put for `Array_1D_Spectrum`?

**Comment by juzen2003** [comment]
Current changes: (updates after 1/19/24 meeting)
- In `PRODUCT_LABEL.xml`
    - Change `hst:moving_target_keyword` to `hst:moving_target_keywords`
    - Add `unit="byte"` to `record_length`, `field_location`, and `field_location` tags
- In `hst_dictionary_support.py`
    - Change `hst:observation_type` value `SPECTROSCOPIC` to `SPECTROGRAPHIC`
    - Change `File_Area` to `File_Area_Ancillary` tag if processing level is "Ancillary"
- In all labels templates, keep `logical_identifier` tag and its content in one line to avoid validator complaining about the unexpected carriage returns

Pending items:
- Get HST dicitonary definition update

**Comment by juzen2003** [comment]
Details of the new updates in https://github.com/SETI/rms-hst-pipeline/issues/56 (comments on 06/17/24)

**Comment by matthewtiscareno** [comment]
> Details of the new updates in #56 (comments on 06/17/24)

Yes, those look good.  So do all the tests pass now?

**Comment by juzen2003** [comment]
06/21/24 updates:
When creating the target label that doesn't exist in https://pds.nasa.gov/data/pds4/context-pds4/target/ , obtain the correct type and name from lid for the target label.

**Pull request merged**